File: Python/merge_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1:
            return list2
        elif not list2:
            return list1
        curr_ptr =  ListNode()
        ret_list = curr_ptr

        while list1 and list2:
            if list1.val < list2.val:
                ret_list.next = list1
                list1 = list1.next
            else:
                ret_list.next = list2
                list2 = list2.next
            ret_list = ret_list.next
        if list1:
            ret_list.next = list1
        if list2:
            ret_list.next = list2
        return curr_ptr.next

File: Cpp/merge_linked_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if (list1 == NULL && list2 == NULL) return NULL;
        if (list1 == NULL) return list2;
        if (list2 == NULL) return list1;

        ListNode* prev_ptr = new ListNode();
        ListNode* curr_ptr = prev_ptr;

        while( list1 != NULL && list2 != NULL){
            if(list1 -> val < list2 -> val){
                curr_ptr -> next = list1;
                list1 = list1 -> next;
            }
            else{
                curr_ptr -> next = list2;
                list2 = list2 -> next;
            }
            curr_ptr =  curr_ptr -> next;
        }
        if (list1 != NULL) curr_ptr ->next = list1;
        else curr_ptr ->next = list2;
        return prev_ptr -> next;
    }
};